Provide an API for controlling and h2c upgrade response stream in Http2MultiplexCodec

bryce-anderson opened the pull request:
Motivation:

Http2MultiplexCodec doesn't currently have an API for using the response
of a h2c upgrade request.

Modifications:

Add a new API to the Http2MultiplexCodecBuilder which allows for setting
an upgrade handler and wire it into the Http2MultiplexCodec
implementation.

Result:

When using the Http2MultiplexCodec with h2c upgrades the upgrade handler
will get added to the Http2StreamChannel which represents the
half-closed (local) response of stream 1. It is then up to the user to
manage the transition from the IO channel pipeline configuration
necessary for making the h2c upgrade request to a form where it can read
the response from the new stream channel.

Fixes #7947.

bryce-anderson commented:
@normanmaurer, @ejona86, this isn't complete but I'd like some early feedback before I start writing tests. 

This isn't as clean as I'd like, and it feels like we're managing some duplicate state. For example, the `DefaultHttp2StreamChannel.outboundClosed` should always mirror the state of the `DefaultHttp2FrameStream.state()`. The `DefaultHttp2FrameStream.id` feels like unnecessary state in that we never set it anything other than `stream.id()` or `-1`, if the stream id hasn't been assigned yet. They didn't look necessary to change in this diff (though I'm happy to do so if folks think it's a good idea), but made me feel like I'm missing something.

ejona86 commented:
By combination of travel, holidays, and meetings, I'm not going to get a chance to look at this until Wednesday.

normanmaurer commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2ClientUpgradeCodecTest.java:
```diff
@@ -31,6 +33,13 @@
+    private class UpgradeStreamHandler extends ChannelInitializer<Channel> {
+        @Override
+        protected void initChannel(Channel ch) throws Exception {
+            return;
```
nit: remove ?

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodecBuilder.java:
```diff
@@ -29,6 +29,7 @@
         extends AbstractHttp2ConnectionHandlerBuilder<Http2MultiplexCodec, Http2MultiplexCodecBuilder> {
 
     final ChannelHandler childHandler;
+    ChannelHandler upgradeStreamHandler = null;
```
you can remove the `= null`

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodecBuilder.java:
```diff
@@ -83,6 +84,13 @@ public Http2MultiplexCodecBuilder gracefulShutdownTimeoutMillis(long gracefulShu
 
+    public Http2MultiplexCodecBuilder withUpgradeStreamHandler(ChannelHandler upgradeStreamHandler) {
+        if (this.isServer())
+            throw new IllegalArgumentException("Server codecs don't utilize an extra handler for the upgrade stream");
```
please enclose with `{}`

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -275,6 +312,7 @@ final void onHttp2StreamWritabilityChanged(ChannelHandlerContext ctx, Http2Frame
     // TODO: This is most likely not the best way to expose this, need to think more about it.
     final Http2StreamChannel newOutboundStream() {
         return new DefaultHttp2StreamChannel(newStream(), true);
+
```
nit: remove

normanmaurer commented:
@bryce-anderson generally speaking I think this change makes sense. I think what you suggested needs more investigation and should be done after this PR is merged.

bryce-anderson commented:
This should be ready for another look.

normanmaurer commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2ClientUpgradeCodecTest.java:
```diff
@@ -43,7 +51,8 @@ public void testUpgradeToHttp2FrameCodec() throws Exception {
     public void testUpgradeToHttp2MultiplexCodec() throws Exception {
-        testUpgrade(Http2MultiplexCodecBuilder.forClient(new HttpInboundHandler()).build());
+        testUpgrade(Http2MultiplexCodecBuilder.forClient(new HttpInboundHandler())
+            .withUpgradeStreamHandler(new UpgradeStreamHandler()).build());
```
You could also just use `ChannelInboundHandlerAdapter` and so not need to declare an extra class .

normanmaurer commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2MultiplexCodecClientUpgradeTest.java:
```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2016 The Netty Project
```
2018

normanmaurer commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2MultiplexCodecClientUpgradeTest.java:
```diff
@@ -0,0 +1,81 @@
+        @Override
+        public void channelActive(ChannelHandlerContext ctx) {
+      ctx.channel().close();
+    }
```
formatting

normanmaurer commented on codec-http2/src/test/java/io/netty/handler/codec/http2/Http2MultiplexCodecClientUpgradeTest.java:
```diff
@@ -0,0 +1,81 @@
+    }
+    }
+
+    private class UpgradeHandler extends ChannelInboundHandlerAdapter {
```
nit: final

normanmaurer commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodecBuilder.java:
```diff
@@ -29,6 +29,7 @@
         extends AbstractHttp2ConnectionHandlerBuilder<Http2MultiplexCodec, Http2MultiplexCodecBuilder> {
 
     final ChannelHandler childHandler;
+    ChannelHandler upgradeStreamHandler;
```
private ?

normanmaurer commented:
@bryce-anderson generally looking great. Left a few more comments

bryce-anderson commented:
@normanmaurer, do you mean in addition to the comments from 6 hours ago? I believe I addressed the comments from a few hours ago. Maybe you have some queued up and just haven't submitted them?

normanmaurer commented:
@carl-mastrangelo @ejona86 can some of you have a look as well...

normanmaurer commented:
@bryce-anderson sorry I did not refresh yet... looking good to me.

carl-mastrangelo commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -236,6 +257,22 @@ final void onHttp2Frame(ChannelHandlerContext ctx, Http2Frame frame) {
+        DefaultHttp2StreamChannel ch = new DefaultHttp2StreamChannel(stream, true);
+        ch.outboundClosed = true;
+
+        // Add our upgrade handler to he channel then register the channel.
```
s/he/the/

mosesn commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java:
```diff
@@ -168,9 +173,25 @@ public int guess() {
+        // Creates the Http2Stream in the Connection.
+        super.onHttpClientUpgrade();
+        // Now make a new FrameStream, set it's underlying Http2Stream, and initialize it.
+        Http2MultiplexCodecStream stream2 = newStream();
```
can we come up with a better name than stream2?

mosesn commented on codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodecBuilder.java:
```diff
@@ -83,6 +84,14 @@ public Http2MultiplexCodecBuilder gracefulShutdownTimeoutMillis(long gracefulShu
 
+    public Http2MultiplexCodecBuilder withUpgradeStreamHandler(ChannelHandler upgradeStreamHandler) {
+        if (this.isServer()) {
+            throw new IllegalArgumentException("Server codecs don't utilize an extra handler for the upgrade stream");
```
ffti: s/utilize/use

normanmaurer commented:
@bryce-anderson thanks for the patch and thanks to @mosesn @carl-mastrangelo @ejona86 for the review.
